src/main/java/org/swisspush/reststorage/s3/S3FileSystemStorage.java

```diff
@@ -0,0 +1,386 @@
+    public static final String S3_PATH_SEPARATOR = "/";
+
+    // use only one file system instance
+    private static S3FileSystem getFileSystem(URI uri) {
+        if (fileSystem == null) {
+            synchronized (S3FileSystem.class) {
+                if (fileSystem == null) {
+                    fileSystem = (S3FileSystem) FileSystems.getFileSystem(uri);
+                }
+            }
+        }
+        return fileSystem;
+    }
+
+    public S3FileSystemStorage(Vertx vertx, RestStorageExceptionFactory exceptionFactory, String rootPath,
+                               String awsS3Region, String awsS3BucketName, String awsS3AccessKeyId, String awsS3SecretAccessKey) {
+        this.vertx = vertx;
+        this.exceptionFactory = exceptionFactory;
+        Objects.requireNonNull(awsS3Region, "Region must not be null");
+        Objects.requireNonNull(awsS3BucketName, "BucketName must not be null");
+        Objects.requireNonNull(awsS3AccessKeyId, "AccessKeyId must not be null");
+        Objects.requireNonNull(awsS3SecretAccessKey, "SecretAccessKey must not be null");
+
+        System.setProperty("aws.region",awsS3Region);
+        System.setProperty("aws.accessKeyId", awsS3AccessKeyId);
+        System.setProperty("aws.secretAccessKey", awsS3SecretAccessKey);
+
+        if (!awsS3BucketName.startsWith("s3:") && !awsS3BucketName.startsWith("s3x:")) {
+            awsS3BucketName = "s3://" + awsS3BucketName;
+        }
+
+        var uri = URI.create(awsS3BucketName);
+        fileSystem = getFileSystem(uri);
+        root = fileSystem.getPath(rootPath);
+
+        try (var pathStream = Files.walk(root)) {
+            pathStream.forEach(System.out::println);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        this.fileSystemDirLister = new S3FileSystemDirLister(vertx, root.toString());
+        // Unify format for simpler work.
+        String tmpRoot;
+        tmpRoot = root.toString();
+
+        // Cache string length of root without trailing slashes
+        int rootLen;
+        for (rootLen = tmpRoot.length() - 1; tmpRoot.charAt(rootLen) == '/'; --rootLen) ;
+        this.rootLen = rootLen;
+    }
+
+    @Override
+    public Optional<Float> getCurrentMemoryUsage() {
+        throw new UnsupportedOperationException("Method 'getCurrentMemoryUsage' not supported in S3FileSystemStorage");
+    }
+
+    @Override
+    public void get(String path, String etag, final int offset, final int count, final Handler<Resource> handler) {
+        final Path fullDirPath = canonicalize(path, true);
+        final Path fullFilePath = canonicalize(path, false);
+        log.debug("GET {}", path);
+        if (Files.isRegularFile(fullFilePath, LinkOption.NOFOLLOW_LINKS)) {
+            log.debug("Open file '{}' ({})", path, fullFilePath);
+            DocumentResource d = new DocumentResource();
+            try {
+                final InputStream inputStream = Files.newInputStream(fullFilePath);
```

Comment:
Why not directly put it in the `try(...) {` statement? Is it because of the async behaviour?
Add a comment if it is because of the async b.

Reply:
done